File: api/src/main/java/org/pmiops/workbench/db/model/Workspace.java
package org.pmiops.workbench.db.model;

import java.sql.Timestamp;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;
import javax.persistence.CascadeType;
import javax.persistence.CollectionTable;
import javax.persistence.Column;
import javax.persistence.ElementCollection;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.Table;
import javax.persistence.Transient;
import javax.persistence.Version;
import org.pmiops.workbench.model.DataAccessLevel;
import org.pmiops.workbench.model.UnderservedPopulationEnum;


@Entity
@Table(name = "workspace")
public class Workspace {

  private String firecloudUuid;

  public static class FirecloudWorkspaceId {
    private final String workspaceNamespace;
    private final String workspaceName;

    public FirecloudWorkspaceId(String workspaceNamespace, String workspaceName) {
      this.workspaceNamespace = workspaceNamespace;
      this.workspaceName = workspaceName;
    }

    public String getWorkspaceNamespace() {
      return workspaceNamespace;
    }

    public String getWorkspaceName() {
      return workspaceName;
    }

    @Override
    public int hashCode() {
      return Objects.hash(workspaceNamespace, workspaceName);
    }

    @Override
    public boolean equals(Object obj) {
      if (!(obj instanceof FirecloudWorkspaceId)) {
        return false;
      }
      FirecloudWorkspaceId that = (FirecloudWorkspaceId) obj;
      return this.workspaceNamespace.equals(that.workspaceNamespace)
          && this.workspaceName.equals(that.workspaceName);
    }
  }

  private long workspaceId;
  private int version;
  private String name;
  private String description;
  private String workspaceNamespace;
  private String firecloudName;
  private DataAccessLevel dataAccessLevel;
  private CdrVersion cdrVersion;
  private User creator;
  private Timestamp creationTime;
  private Timestamp lastModifiedTime;
  private Set<Cohort> cohorts = new HashSet<Cohort>();

  private boolean diseaseFocusedResearch;
  private String diseaseOfFocus;
  private boolean methodsDevelopment;
  private boolean controlSet;
  private boolean aggregateAnalysis;
  private boolean ancestry;
  private boolean commercialPurpose;
  private boolean population;
  private String populationOfFocus;
  private String additionalNotes;
  private boolean containsUnderservedPopulation;
  private Set<UnderservedPopulationEnum> underservedPopulationSet =
      new HashSet<UnderservedPopulationEnum>();

  private Boolean reviewRequested;
  private Boolean approved;
  private Timestamp timeRequested;

  private Set<WorkspaceUserRole> usersWithAccess = new HashSet<WorkspaceUserRole>();

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  @Column(name = "workspace_id")
  public long getWorkspaceId() {
    return workspaceId;
  }

  public void setWorkspaceId(long workspaceId) {
    this.workspaceId = workspaceId;
  }

  @Version
  @Column(name = "version")
  public int getVersion() {
    return version;
  }

  public void setVersion(int version) { this.version = version; }

  @Column(name = "name")
  public String getName() {
    return name;
  }

  public void setName(String name) {
    this.name = name;
  }

  @Column(name = "description")
  public String getDescription() {
    return description;
  }

  public void setDescription(String description) {
    this.description = description;
  }

  @Column(name = "workspace_namespace")
  public String getWorkspaceNamespace() {
    return workspaceNamespace;
  }

  public void setWorkspaceNamespace(String workspaceNamespace) {
    this.workspaceNamespace = workspaceNamespace;
  }


  @Column(name = "firecloud_name")
  public String getFirecloudName() {
    return firecloudName;
  }

  public void setFirecloudName(String firecloudName) {
    this.firecloudName = firecloudName;
  }

  @Column(name = "data_access_level")
  public DataAccessLevel getDataAccessLevel() {
    return dataAccessLevel;
  }

  public void setDataAccessLevel(DataAccessLevel dataAccessLevel) {
    this.dataAccessLevel = dataAccessLevel;
  }

  @ManyToOne
  @JoinColumn(name = "cdr_version_id")
  public CdrVersion getCdrVersion() {
    return cdrVersion;
  }

  public void setCdrVersion(CdrVersion cdrVersion) {
    this.cdrVersion = cdrVersion;
  }

  @ManyToOne
  @JoinColumn(name = "creator_id")
  public User getCreator() {
    return creator;
  }

  public void setCreator(User creator) {
    this.creator = creator;
  }

  @Column(name = "creation_time")
  public Timestamp getCreationTime() {
    return creationTime;
  }

  public void setCreationTime(Timestamp creationTime) {
    this.creationTime = creationTime;
  }

  @Column(name = "last_modified_time")
  public Timestamp getLastModifiedTime() {
    return lastModifiedTime;
  }

  public void setLastModifiedTime(Timestamp lastModifiedTime) {
    this.lastModifiedTime = lastModifiedTime;
  }

  @Column(name = "rp_disease_focused_research")
  public boolean getDiseaseFocusedResearch() {
    return this.diseaseFocusedResearch;
  }

  public void setDiseaseFocusedResearch(boolean diseaseFocusedResearch) {
    this.diseaseFocusedResearch = diseaseFocusedResearch;
  }

  @Column(name = "rp_disease_of_focus")
  public String getDiseaseOfFocus() {
    return this.diseaseOfFocus;
  }

  public void setDiseaseOfFocus(String diseaseOfFocus) {
    this.diseaseOfFocus = diseaseOfFocus;
  }

  @Column(name = "rp_methods_development")
  public boolean getMethodsDevelopment() {
    return this.methodsDevelopment;
  }

  public void setMethodsDevelopment(boolean methodsDevelopment) {
    this.methodsDevelopment = methodsDevelopment;
  }

  @Column(name = "rp_control_set")
  public boolean getControlSet() {
    return this.controlSet;
  }

  public void setControlSet(boolean controlSet) {
    this.controlSet = controlSet;
  }

  @Column(name = "rp_aggregate_analysis")
  public boolean getAggregateAnalysis() {
    return this.aggregateAnalysis;
  }

  public void setAggregateAnalysis(boolean aggregateAnalysis) {
    this.aggregateAnalysis = aggregateAnalysis;
  }

  @Column(name = "rp_ancestry")
  public boolean getAncestry() {
    return this.ancestry;
  }

  public void setAncestry(boolean ancestry) {
    this.ancestry = ancestry;
  }

  @Column(name = "rp_commercial_purpose")
  public boolean getCommercialPurpose() {
    return this.commercialPurpose;
  }

  public void setCommercialPurpose(boolean commercialPurpose) {
    this.commercialPurpose = commercialPurpose;
  }

  @Column(name = "rp_population")
  public boolean getPopulation() {
    return this.population;
  }

  public void setPopulation(boolean population) {
    this.population = population;
  }

  @Column(name = "rp_population_of_focus")
  public String getPopulationOfFocus() {
    return this.populationOfFocus;
  }

  public void setPopulationOfFocus(String populationOfFocus) {
    this.populationOfFocus = populationOfFocus;
  }

  @Column(name = "rp_additional_notes")
  public String getAdditionalNotes() {
    return this.additionalNotes;
  }

  public void setAdditionalNotes(String additionalNotes) {
    this.additionalNotes = additionalNotes;
  }

  @Column(name = "rp_contains_underserved_population")
  public Boolean getContainsUnderservedPopulation() {
    return this.containsUnderservedPopulation;
  }

  public void setContainsUnderservedPopulation(Boolean containsUnderservedPopulation) {
    this.containsUnderservedPopulation = containsUnderservedPopulation;
  }

  @ElementCollection(fetch = FetchType.LAZY)
  @CollectionTable(name = "underserved_populations", joinColumns = @JoinColumn(name = "workspace_id"))
  @Column(name = "underserved_population")
  public Set<UnderservedPopulationEnum> getUnderservedPopulationSet() {
    return underservedPopulationSet;
  }

  public void setUnderservedPopulationSet(Set<UnderservedPopulationEnum> newUnderservedPopulationSet) {
    this.underservedPopulationSet = newUnderservedPopulationSet;
  }

  @Column(name = "rp_review_requested")
  public Boolean getReviewRequested() {
    return this.reviewRequested;
  }

  public void setReviewRequested(Boolean reviewRequested) {
    if (reviewRequested != null && reviewRequested && this.timeRequested == null) {
      this.timeRequested = new Timestamp(System.currentTimeMillis());
    }
    this.reviewRequested = reviewRequested;
  }

  @Column(name = "rp_approved")
  public Boolean getApproved() {
    return this.approved;
  }

  public void setApproved(Boolean approved) {
    this.approved = approved;
  }

  @Column(name = "rp_time_requested")
  public Timestamp getTimeRequested() {
    return this.timeRequested;
  }

  public void setTimeRequested(Timestamp timeRequested) {
    this.timeRequested = timeRequested;
  }

  @OneToMany(mappedBy = "workspaceId", orphanRemoval = true, cascade = CascadeType.ALL)
  public Set<Cohort> getCohorts() {
    return cohorts;
  }

  public void setCohorts(Set<Cohort> cohorts) {
    this.cohorts = cohorts;
  }

  public void addCohort(Cohort cohort) {
    this.cohorts.add(cohort);
  }

  @Transient
  public FirecloudWorkspaceId getFirecloudWorkspaceId() {
    return new FirecloudWorkspaceId(workspaceNamespace, firecloudName);
  }

  @OneToMany(fetch = FetchType.EAGER, mappedBy = "workspace", orphanRemoval = true, cascade = CascadeType.ALL)
  public Set<WorkspaceUserRole> getWorkspaceUserRoles() {
    return usersWithAccess;
  }

  @Column(name = "firecloud_uuid")
  public String getFirecloudUuid() {return this.firecloudUuid; }

  public void setFirecloudUuid(String firecloudUuid) {this.firecloudUuid = firecloudUuid;}

  /**
   * Necessary for Spring initialization of the object.
   * Not actually supported because it won't delete old entries.
   */
  public void setWorkspaceUserRoles(Set<WorkspaceUserRole> userRoles) {
    this.usersWithAccess = userRoles;
  }

  public void addWorkspaceUserRole(WorkspaceUserRole userRole) {
    this.usersWithAccess.add(userRole);
  }
}

File: api/src/main/java/org/pmiops/workbench/db/model/Workspace.kt
package org.pmiops.workbench.db.model

import java.sql.Timestamp
import java.util.HashSet
import java.util.Objects
import java.util.stream.Collectors
import javax.persistence.CascadeType
import javax.persistence.CollectionTable
import javax.persistence.Column
import javax.persistence.ElementCollection
import javax.persistence.Entity
import javax.persistence.FetchType
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.Table
import javax.persistence.Transient
import javax.persistence.Version
import org.pmiops.workbench.model.BillingAccountType
import org.pmiops.workbench.model.BillingStatus
import org.pmiops.workbench.model.DataAccessLevel
import org.pmiops.workbench.model.SpecificPopulationEnum
import org.pmiops.workbench.model.WorkspaceActiveStatus

@Entity
@Table(name = "workspace")
class Workspace {
    @get:Column(name = "firecloud_uuid")
    var firecloudUuid: String? = null

    @get:Id
    @get:GeneratedValue(strategy = GenerationType.IDENTITY)
    @get:Column(name = "workspace_id")
    var workspaceId: Long = 0
    @get:Version
    @get:Column(name = "version")
    var version: Int = 0
    @get:Column(name = "name")
    var name: String? = null
    @get:Column(name = "workspace_namespace")
    var workspaceNamespace: String? = null
    @get:Column(name = "firecloud_name")
    var firecloudName: String? = null
    @get:Column(name = "data_access_level")
    var dataAccessLevel: Short? = null
    @get:ManyToOne
    @get:JoinColumn(name = "cdr_version_id")
    var cdrVersion: CdrVersion? = null
    @get:ManyToOne
    @get:JoinColumn(name = "creator_id")
    var creator: User? = null
    @get:Column(name = "creation_time")
    var creationTime: Timestamp? = null
    @get:Column(name = "last_modified_time")
    var lastModifiedTime: Timestamp? = null
    @get:Column(name = "last_accessed_time")
    var lastAccessedTime: Timestamp? = null
    private var cohorts: MutableSet<Cohort> = HashSet()
    @get:OneToMany(mappedBy = "workspaceId", orphanRemoval = true, cascade = [CascadeType.ALL])
    var conceptSets: Set<ConceptSet> = HashSet()
    private var dataSets: MutableSet<DataSet> = HashSet()
    @get:Column(name = "active_status")
    private var activeStatus: Short? = null
    private var billingMigrationStatus = StorageEnums.billingMigrationStatusToStorage(BillingMigrationStatus.OLD)
    @get:Column(name = "published")
    var published: Boolean = false

    @get:Column(name = "rp_disease_focused_research")
    var diseaseFocusedResearch: Boolean = false
    @get:Column(name = "rp_disease_of_focus")
    var diseaseOfFocus: String? = null
    @get:Column(name = "rp_methods_development")
    var methodsDevelopment: Boolean = false
    @get:Column(name = "rp_control_set")
    var controlSet: Boolean = false
    @get:Column(name = "rp_ancestry")
    var ancestry: Boolean = false
    @get:Column(name = "rp_commercial_purpose")
    var commercialPurpose: Boolean = false
    @get:Column(name = "rp_population")
    var population: Boolean = false
    @get:ElementCollection(fetch = FetchType.LAZY)
    @get:CollectionTable(name = "specific_populations", joinColumns = [JoinColumn(name = "workspace_id")])
    @get:Column(name = "specific_population")
    var populationDetails: Set<Short>? = HashSet()
    @get:Column(name = "rp_social_behavioral")
    var socialBehavioral: Boolean = false
    @get:Column(name = "rp_population_health")
    var populationHealth: Boolean = false
    @get:Column(name = "rp_educational")
    var educational: Boolean = false
    @get:Column(name = "rp_drug_development")
    var drugDevelopment: Boolean = false
    @get:Column(name = "rp_other_purpose")
    var otherPurpose: Boolean = false
    @get:Column(name = "rp_other_purpose_details")
    var otherPurposeDetails: String? = null
    @get:Column(name = "rp_other_population_details")
    var otherPopulationDetails: String? = null
    @get:Column(name = "rp_additional_notes")
    var additionalNotes: String? = null
    @get:Column(name = "rp_reason_for_all_of_us")
    var reasonForAllOfUs: String? = null
    @get:Column(name = "rp_intended_study")
    var intendedStudy: String? = null
    @get:Column(name = "rp_anticipated_findings")
    var anticipatedFindings: String? = null

    @get:Column(name = "rp_review_requested")
    var reviewRequested: Boolean? = null
        set(reviewRequested) {
            if (reviewRequested != null && reviewRequested && this.timeRequested == null) {
                this.timeRequested = Timestamp(System.currentTimeMillis())
            }
            field = reviewRequested
        }
    @get:Column(name = "rp_approved")
    var approved: Boolean? = null
    @get:Column(name = "rp_time_requested")
    var timeRequested: Timestamp? = null
    private var billingStatus = StorageEnums.billingStatusToStorage(BillingStatus.ACTIVE)
    private var billingAccountType = StorageEnums.billingAccountTypeToStorage(BillingAccountType.FREE_TIER)

    var dataAccessLevelEnum: DataAccessLevel
        @Transient
        get() = CommonStorageEnums.dataAccessLevelFromStorage(dataAccessLevel)
        set(dataAccessLevel) {
            dataAccessLevel = CommonStorageEnums.dataAccessLevelToStorage(dataAccessLevel)
        }

    var specificPopulationsEnum: Set<SpecificPopulationEnum>?
        @Transient
        get() {
            val from = populationDetails ?: return null
            return from.stream()
                    .map(Function<Short, R> { StorageEnums.specificPopulationFromStorage(it) })
                    .collect(Collectors.toSet<Any>())
        }
        set(newPopulationDetails) {
            populationDetails = newPopulationDetails.stream()
                    .map<Short>(Function<SpecificPopulationEnum, Short> { StorageEnums.specificPopulationToStorage(it) })
                    .collect<Set<Short>, Any>(Collectors.toSet())
        }

    val firecloudWorkspaceId: FirecloudWorkspaceId
        @Transient
        get() = FirecloudWorkspaceId(workspaceNamespace, firecloudName)

    var workspaceActiveStatusEnum: WorkspaceActiveStatus
        @Transient
        get() = StorageEnums.workspaceActiveStatusFromStorage(activeStatus)
        set(activeStatus) {
            activeStatus = StorageEnums.workspaceActiveStatusToStorage(activeStatus)
        }

    var billingMigrationStatusEnum: BillingMigrationStatus
        @Transient
        get() = StorageEnums.billingMigrationStatusFromStorage(billingMigrationStatus)
        set(status) {
            this.billingMigrationStatus = StorageEnums.billingMigrationStatusToStorage(status)
        }

    class FirecloudWorkspaceId(val workspaceNamespace: String, val workspaceName: String) {

        override fun hashCode(): Int {
            return Objects.hash(workspaceNamespace, workspaceName)
        }

        override fun equals(obj: Any?): Boolean {
            if (obj !is FirecloudWorkspaceId) {
                return false
            }
            val that = obj as FirecloudWorkspaceId?
            return this.workspaceNamespace == that!!.workspaceNamespace && this.workspaceName == that.workspaceName
        }
    }

    // see https://precisionmedicineinitiative.atlassian.net/browse/RW-2705
    enum class BillingMigrationStatus {
        OLD, // pre-1PPW; this billing project may be associated with multiple workspaces
        NEW, // a One Project Per Workspace (1PPW) billing project
        MIGRATED // a pre-1PPW billing project which has been cloned and is now ready to be deleted
    }

    init {
        workspaceActiveStatusEnum = WorkspaceActiveStatus.ACTIVE
    }

    @OneToMany(mappedBy = "workspaceId", orphanRemoval = true, cascade = [CascadeType.ALL])
    fun getCohorts(): Set<Cohort> {
        return cohorts
    }

    fun setCohorts(cohorts: MutableSet<Cohort>) {
        this.cohorts = cohorts
    }

    fun addCohort(cohort: Cohort) {
        this.cohorts.add(cohort)
    }

    @OneToMany(mappedBy = "workspaceId", orphanRemoval = true, cascade = [CascadeType.ALL])
    fun getDataSets(): Set<DataSet> {
        return dataSets
    }

    fun setDataSets(dataSets: MutableSet<DataSet>) {
        this.dataSets = dataSets
    }

    fun addDataSet(dataSet: DataSet) {
        this.dataSets.add(dataSet)
    }

    @Column(name = "billing_migration_status")
    private fun getBillingMigrationStatus(): Short {
        return this.billingMigrationStatus!!
    }

    private fun setBillingMigrationStatus(s: Short) {
        this.billingMigrationStatus = s
    }

    @Column(name = "billing_status")
    fun getBillingStatus(): BillingStatus {
        return StorageEnums.billingStatusFromStorage(billingStatus)
    }

    fun setBillingStatus(billingStatus: BillingStatus) {
        this.billingStatus = StorageEnums.billingStatusToStorage(billingStatus)
    }

    @Column(name = "billing_account_type")
    fun getBillingAccountType(): BillingAccountType {
        return StorageEnums.billingAccountTypeFromStorage(billingAccountType)
    }

    fun setBillingAccountType(billingAccountType: BillingAccountType) {
        this.billingAccountType = StorageEnums.billingAccountTypeToStorage(billingAccountType)
    }
}
